src/inmanta/agent/forking_executor.py

```diff
@@ -710,13 +753,40 @@ async def stop_for_agent(self, agent_name: str) -> list[MPExecutor]:
         await asyncio.gather(*(child.stop() for child in children))
         return children
 
-    async def cleanup_inactive_executors(self) -> None:
-        now = datetime.datetime.now().astimezone()
+    def start_periodic_executor_cleanup(self) -> None:
+        async def cleanup_inactive_executors() -> None:
+            """
+            This task cleans up idle executors and reschedules itself
+            """
 
-        for _executor in self.executor_map.values():
-            if now - _executor.connection.last_used_at > timedelta(seconds=self.executor_retention_time):
-                LOGGER.info(
-                    f"Stopping executor {_executor.executor_id.identity()} because it was inactive for longer "
-                    f"than the configured allowed idling time."
-                )
-                await _executor.stop()
+            now = datetime.datetime.now().astimezone()
+            reschedule_interval: float = self.executor_retention_time
+            for _executor in self.executor_map.values():
+                if _executor.can_be_cleaned_up(self.executor_retention_time):
+                    LOGGER.debug(
+                        "Stopping executor %s because it "
+                        "was inactive for %d s, which is longer then the retention time of %d s.",
+                        _executor.executor_id.identity(),
+                        _executor.connection.get_idle_time().total_seconds(),
+                        self.executor_retention_time,
+                    )
+                    async with self._locks.get(_executor.executor_id.identity()):
+                        try:
+                            await _executor.stop()
```

Comment:
If there's any chance this might hang we should take into account the difference between `now` and the time when we reschedule at the end of the method. But I get the impression that this should always return quickly?

Reply:
Added mechanism to take into account the difference between start and end of cleanup